jupyterhub_singleuser_profiles/profiles.py

```diff
@@ -101,21 +120,62 @@ def write_config_map(self, config_map_name, key_name, data):
       else:
         raise
 
+  def write_secret(self, secret_name, key_name, data):
+    secret = V1Secret()
+    secret.metadata = V1ObjectMeta(name=secret_name, labels={'app': 'jupyterhub'})
+    secret.string_data = {key_name: yaml.dump(data, default_flow_style=False)} #stringData instead of data here to make kubernetes parse this as a string without base64 encoding
+    try:
+      api_response = self.api_client.replace_namespaced_secret(secret_name, self.namespace, secret)
+    except ApiException as e:
+      if e.status == 404:
+        try:
+          api_response = self.api_client.create_namespaced_secret(self.namespace, secret)
+        except ApiException as e:
+          _LOGGER.error("Exception when calling CoreV1Api->create_namespaced_secret: %s\n" % e)
+      else:
+        raise
+
   def update_user_profile_cm(self, username, data={}):
     
     user_cm = self.get_user_profile_cm(username)
+    user_cm_env = []
+    user_secret = self.get_user_secret(username)
+    user_secret['env'] = []
     cm_name = _USER_CONFIG_MAP_TEMPLATE % escape(username)
     cm_key_name = "profile"
     for key, value in data.items():
       user_cm[key] = value
+    for var in user_cm['env']:
+      if var['type'] == "password":
+        user_secret['env'].append(var) 
+      else:
+        user_cm_env.append(var)
+    user_cm['env'] = user_cm_env
+    self.write_secret(cm_name, cm_key_name, user_secret) #not a typo, secret uses the same name as user cm
     self.write_config_map(cm_name, cm_key_name, user_cm)
 
   def get_user_profile_cm(self, username):
     cm = self.read_config_map(_USER_CONFIG_MAP_TEMPLATE % escape(username), "profile")
+    secret = self.get_user_secret(username)
     if cm == {}:
-      return _DEFAULT_USER_CM
+      cm = _DEFAULT_USER_CM
+    for sec in secret['env']:
```

Comment:
You will need to check if the env already exists here, for the case where admin adds some CM env or secret env manually and it collides with one that already exists - see

![Screenshot from 2021-03-01 16-09-04](https://user-images.githubusercontent.com/4759808/109516443-7f135580-7aa8-11eb-9a08-58890a4d4a44.png)